api/src/main/java/io/smallrye/reactive/messaging/Messages.java

```diff
@@ -121,4 +174,103 @@ private static Metadata merge(Metadata first, Metadata second) {
         return result;
     }
 
+    /**
+     * The message chain builder allows chaining message and configure metadata propagation.
+     * By default, all the metadata from the given message are copied into the chained messages.
+     */
+    public static class MessageChainBuilder {
+        private final Message<?> input;
+        private Metadata metadata;
+
+        private MessageChainBuilder(Message<?> message) {
+            this.input = message;
+            this.metadata = message.getMetadata().copy();
+        }
+
+        /**
+         * Do not copy any metadata from the initial message to the chained message.
+         *
+         * @return the current {@link MessageChainBuilder}
+         */
+        @CheckReturnValue
+        public MessageChainBuilder withoutMetadata() {
+            this.metadata = Metadata.empty();
+            return this;
+        }
+
+        /**
+         * Copy the given metadata of the given classes from the initial message to the chained message, if the initial
+         * message does not include a metadata object of the given class.
+         *
+         * In general, this method must be used after {@link #withoutMetadata()}.
+         *
+         * @return the current {@link MessageChainBuilder}
+         */
+        @CheckReturnValue
+        public MessageChainBuilder withMetadata(Class<?>... mc) {
+            for (Class<?> clazz : mc) {
+                Optional<?> o = input.getMetadata().get(clazz);
+                o.ifPresent(value -> this.metadata = metadata.with(value));
+            }
+            return this;
+        }
+
+        /**
+         * Do not the given metadata of the given classes from the initial message to the chained message, if the initial
+         * message does not include a metadata object of the given class.
+         *
+         * @return the current {@link MessageChainBuilder}
+         */
+        @CheckReturnValue
+        public MessageChainBuilder withoutMetadata(Class<?>... mc) {
+            for (Class<?> clazz : mc) {
+                this.metadata = this.metadata.without(clazz);
+            }
+            return this;
+        }
+
+        /**
+         * Passed the chained messages.
+         * The messages are not modified, but should not be used afterward, and should be replaced by the messages contained
+         * in the returned list.
+         * This method preserve the order. So, the first message corresponds to the first message in the returned list.
+         * The message from the returned list have the necessary logic to chain the ack/nack signals and the copied metadata.
+         *
+         * @param messages the chained messages, must not be empty, must not be null, must not contain null
+         * @return the list of modified messages
+         */
+        public List<Message<?>> with(Message<?>... messages) {
```

Comment:
We may think about wrapping the return into a marker type to integrate with `@Outgoings`. 

If multiple messages are returned we redirect them by order of declaration. If a single message is returned it'd be considered as broadcast.